### ISSUE TITLE
Difference about adding `eval` or not

### ISSUE THREAD

**Issue by chunyang-wen** [ml-explore/mlx-examples#268]
In [MNIST main.py](https://github.com/ml-explore/mlx-examples/blob/6e6eff326e9a615f3d9870706eb831afb32dc9e8/mnist/main.py#L66), there are function calls of `eval`.

```python
mx.eval(model.parameters()) # L66

mx.eval(model.parameters(), optimizer.state) # L76
```

But I can also remove the `eval` and everything works fine.

From [the document](https://ml-explore.github.io/mlx/build/html/usage/quick_start.html#quick-start-guide)

> Operations in MLX are lazy. The outputs of MLX operations are not computed until they are needed. To force an array to be evaluated use [eval()](https://ml-explore.github.io/mlx/build/html/python/_autosummary/mlx.core.eval.html#mlx.core.eval).

So what is the difference here if I remove the two `eval` calls in main.py?

- Performance
- Correctness

**Comment by angeloskath** [comment]
When doing `eval` the arrays are computed at that time. Otherwise they are computed when needed. In the MNIST example, when removing eval, the arrays are computed when printing the loss for instance. I will add a small example to help me explain what happens:

```python
model = MyAwesomeModel()
# At this point model.parameters() is not evaluated.
# We simply have a lot of arrays that when evaluated
# will result in randomly initialized weights.

for iteration, batch in enumerate(my_dataset):
    loss, grad = loss_and_grad_fn(model, batch)
    # Still nothing evaluated. Loss is the forward pass and grad are
    # arrays that when evaluated will run the backward pass.
    
    optimizer.update(model, grad)
    # Still nothing evaluated. The parameters now are arrays that
    # when evaluated will do 1 gradient update.
    
    if ((iteration + 1) % 10) == 0:
        print(loss)
        # Now this will evaluate the 10th loss value. In order to do that, it will
        # do 10 forward passes and 9 backward passes. It will also initialize
        # the random weights in the beginning.
```

So I hope the example above is somewhat illuminating. In theory we don't care about when an array is evaluated. However it may affect the performance as we may end up evaluating huge huge graphs in the background if we evaluate once every 10 steps for instance. Imagine a 10 layer network evaluated every 10 steps. The computation graph that exists in the background may be several thousand operations. Just processing this graph in order to evaluate it may have some overhead.

Another reason it may affect performance is the unnecessary keeping of arrays in memory because of partial evaluation. As an example image the following situation:

```python
x = mx.random.rand(shape=(1024, 1024))
y = x.sum(0)
z = x.sum(1)
# So far nothing is evaluated (not even x)
print(y)
# Now y is obviously evaluated and in order to do that we evaluated x.
# However, z is not evaluated and as a result we need to keep x in memory
# because it is needed to evaluate z.
#
# Evaluating both y and z makes the needed memory 2048 floats rather than 1M floats.
```

**Comment by LeonEricsson** [comment]
@angeloskath what motivated the choice of lazy evaluation? Do you have a reference for understanding the benefits of this, particularly for ML?

**Comment by LeonEricsson** [comment]
> I honestly think that the lazy evaluation is one of the coolest things about MLX. It allows us to do some things simply and efficiently that would be cumbersome otherwise.
> 
> It is for sure not the only one, but one example that everyone benefits immediately from is initializing big models and then loading their weights without actually initializing them randomly. To achieve the same in PT one would need to initialize the model on the "meta" device and then load the weights on the correct device and then update the model. The simple pattern of `model = MyModel()`, `model.load_state_dict(torch.load("path", map_location="cpu"))` and `model.to(device)` can be quite slow in comparison.
> 
> Importantly, it also allows us to have behaviors similar to compiled/static graph frameworks like Jax without imposing a static graph. `mx.simplify` is a good first example. We quickly go through the graph and eliminate some redundant computation but do not require the user to "jit" anything. Another one is our function transformations. You never deal with "symbolic" arrays in MLX function transformations. This means you can print, put a breakpoint have loops and ifs that depend on the values of arrays (like in PT) all being into a function transformation.
> 
> The main thing is that it should be completely transparent to the user except for the need to call `mx.eval` once in the training loop.

Damn, that's really cool thanks for enlightening me. Honestly **lazy evaluation** deserves more of a shoutout in the docs, perhaps with examples like you've given here. I think a lot of people don't really understand the benefits or how to deal with it

**Comment by chunyang-wen** [comment]
Thanks for the detailed explanation. I have the feeling that no `eval` related operations called will introduce a larger graph when we keep adding operations. So it seems we need to be aware of when we should add related eval operations.
